keps/sig-instrumentation/1748-pod-resource-metrics/README.md

```diff
@@ -0,0 +1,431 @@
+
+Note on naming: The Prometheus convention for a metric is to allow the sum of a metric to make sense. However, in the case of Kubernetes resources we are attempting to make discoverability of the reported metrics (which are unbounded in number as they are extensible) a key goal for admins. Most admins will understand CPU and memory usage, but would have to resort to metrics name wildcard regexes to find the other values if we give each resource its own metrics name. Also, because Kube extension resources have characters disallowed in metrics names, admins would continually have to transform `myresource.io/foo-bar` into `myresource_io_foo_bar` in order to locate names which places an undue burden on the user. So for the proposed metrics and `kube_node_status_allocatable` we instead place all types resources under the same metrics name with a unit and resource label.  For this metric, the sum of series is only meaningful when filtering on a resource label.
+
+* `kube_pod_resource_requests` contains series for the consumed requested resources for a pod at a given time.
+  * The value is the quantity of the request in `spec.containers.resources.requests`.
+  * It has the following labels:
+    * `pod` - the name of the pod
+    * `namespace` - the namespace of the pod
+    * `node` - the node the resource is scheduled to, or empty if not yet scheduled
+    * `resource` - the name of the resource as described in `spec.containers.resources.requests`
+    * `unit` - the units of the resource type as inferred from the resource name in `spec.containers.requests`, such as `bytes`, `cores`. Empty if this is a unitless resource or if it is an extension resource that does not have a detectable unit.
+* `kube_pod_resource_limits` contains series for the consumed limit resources for a pod at a given time.
+  * The value is the quantity of the request in `spec.containers.resources.limits`.
+  * It has the following labels:
+    * `pod` - the name of the pod
+    * `namespace` - the namespace of the pod
+    * `node` - the node the resource is scheduled to, or empty if not yet scheduled
+    * `resource` - the name of the resource as described in `spec.containers.resources.limits`
+    * `unit` - the units of the resource type as inferred from the resource name in `spec.containers.limits`, such as `bytes`, `cores`. Empty if this is a unitless resource.
+
+The following metrics will be reported for pods that are in the `Running` phase and follow the same rules as above, but are not reported when the pod is `Pending` and can be used to calculate live resources. They all share the form `kube_running_pod_resource_(requests|limits)[_<resource>_<unit>]`.
```

Comment:
You might want to place this metric in the kubelet (unless we place everything in the kube-apiserver).

We don't really keep track of the Pod phase in the scheduler, because the scheduler doesn't care about running vs non-running.

Reply:
The metric can’t be in the Kubelet because need to report it for pods that aren’t scheduled yet.  Practically speaking KSM could own this if the logic is shared but I do not like letting components off the hook to represent their worldview.  A more sophisticated / custom scheduler could very well have a more powerful resource mode that *should* alter the metrics returned here, which was the original argument to tie this to the component that mediated between request (api) and satisfying the request (kubelet)

Reply:
> The metric can’t be in the Kubelet because need to report it for pods that aren’t scheduled yet.

But these are Running Pods, they are already scheduled.

Reply:
Specifically:

* we should locate metrics close to the responsible party (scheduler bridges request and intent)
* flexibility for custom interpretations in the future
* components already have a pod cache
* access to the standard helpers for this to avoid code mistakes (staging helps but when someone merges a new complication like pod overhead we shouldn’t forget / fail to update downstreams)
* scheduler owns communicating why a scheduling decision is not made, and the visibility of those resources is a facet of that

Reply:
> we should locate metrics close to the responsible party (scheduler bridges request and realization fo request)

scheduler is not responsible for Running Pods :)

> access to the standard helpers for this to avoid code mistakes

kubelet does the requests calculation as well.

Reply:
I'm confused on this as well, when would the scheduler report this information for pods that have been scheduled? It seems like the scheduler would only be able to report for `Pending` pods

Reply:
Oh, maybe this question is related to https://github.com/kubernetes/enhancements/pull/1916/files#r499975216?

Reply:
Scheduler is responsible for transitioning pending pods to runnable pods based on resources.  There is no other component more central to that transition.

Reply:
Not at all. That's the kubelet.

Scheduler is responsible for setting Pod.spec.nodeName, nothing more

Reply:
Some of the confusion I think is occurring is the distinction between the idealized model of resources (the Kubernetes model) and the implementation details of the scheduler.  In general today the metrics reported come from the idealized model - none of the internal implementation details of the scheduler leak.  The component reporting the metrics does not take an undue burden by exposing these, but the key characteristic is ensuring they have fidelity to the idealized model.  The scheduler has to have access to the idealized model today in order to schedule, and any internal representation it chooses still has to honor the model (and efficiency gains on the internal representation would carry over to the metrics source).

Architecturally it’s the most appropriate component on the dimensions I mentioned above, and in practice most clusters leveraging any of the instrumentation we expose will want to answer the questions described in the KEP.